File: packages/sandboxd/src/startup_diagnostics/operation_records.rs
//! Asynchronous operation-record delivery for startup diagnostics.
//!
//! Initialization paths enqueue lifecycle records here so network delivery can
//! retry briefly without blocking the main startup flow indefinitely.

use std::time::Instant;

use serde_json::{Map, Value, json};
use tokio::sync::mpsc;

use crate::startup_diagnostics::{
    LIFECYCLE_OPERATION_RECORD_SEND_RETRY_INTERVAL, LIFECYCLE_OPERATION_RECORD_SEND_TIMEOUT,
    StartupOperation,
};
use crate::tunnel::session::OperationStreamMessage;

pub(super) fn send_lifecycle_operation_record_with_timeout(
    sender: mpsc::Sender<OperationStreamMessage>,
    mut operation_record: OperationStreamMessage,
) {
    let deadline = Instant::now() + LIFECYCLE_OPERATION_RECORD_SEND_TIMEOUT;
    loop {
        match sender.try_send(operation_record) {
            Ok(()) => return,
            Err(mpsc::error::TrySendError::Closed(_)) => {
                eprintln!("sandboxd dropped lifecycle operation record because stream is closed");
                return;
            }
            Err(mpsc::error::TrySendError::Full(returned_record)) => {
                if Instant::now() >= deadline {
                    eprintln!("sandboxd dropped lifecycle operation record after send timeout");
                    return;
                }
                operation_record = returned_record;
                std::thread::sleep(LIFECYCLE_OPERATION_RECORD_SEND_RETRY_INTERVAL);
            }
        }
    }
}

pub(super) fn operation_record_line(
    operation: StartupOperation,
    observed_at: String,
    event: &str,
    payload: &Value,
) -> Result<Option<String>, String> {
    let record = if event == started_event_name(operation) || event == failed_event_name(operation)
    {
        return Ok(None);
    } else if event == phase_started_event_name(operation) {
        let Some(phase) = payload.get("phase").and_then(Value::as_str) else {
            return Ok(None);
        };
        let Some(phase) = operation_lifecycle_phase(phase) else {
            return Ok(None);
        };
        json!({
            "kind": "lifecycle",
            "observedAt": observed_at,
            "phase": phase,
            "status": "started",
            "source": "sandboxd",
            "message": format!("{phase} started"),
            "attributes": lifecycle_attributes(payload)
        })
    } else if event == phase_completed_event_name(operation) {
        let Some(phase) = payload.get("phase").and_then(Value::as_str) else {
            return Ok(None);
        };
        let Some(phase) = operation_lifecycle_phase(phase) else {
            return Ok(None);
        };
        json!({
            "kind": "lifecycle",
            "observedAt": observed_at,
            "phase": phase,
            "status": "completed",
            "source": "sandboxd",
            "message": format!("{phase} completed"),
            "attributes": lifecycle_attributes(payload)
        })
    } else if event == phase_failed_event_name(operation) {
        let Some(phase) = payload.get("phase").and_then(Value::as_str) else {
            return Ok(None);
        };
        let Some(phase) = operation_lifecycle_phase(phase) else {
            return Ok(None);
        };
        json!({
            "kind": "lifecycle",
            "observedAt": observed_at,
            "phase": phase,
            "status": "failed",
            "source": "sandboxd",
            "message": format!("{phase} failed"),
            "attributes": lifecycle_attributes(payload)
        })
    } else if event == transcript_event_name(operation) {
        json!({
            "kind": "transcript",
            "observedAt": observed_at,
            "phase": payload
                .get("phase")
                .and_then(Value::as_str)
                .and_then(operation_lifecycle_phase),
            "source": "sandboxd",
            "stream": payload.get("stream").and_then(Value::as_str).unwrap_or("system"),
            "payloadBase64": payload
                .get("payloadBase64")
                .and_then(Value::as_str)
                .unwrap_or("")
        })
    } else {
        return Ok(None);
    };

    let mut line = serde_json::to_string(&record)
        .map_err(|error| format!("failed to serialize operation record: {error}"))?;
    line.push('\n');
    Ok(Some(line))
}

pub(super) fn operation_lifecycle_phase(phase: &str) -> Option<&'static str> {
    match phase {
        "apply_git_identity" => Some("git_identity"),
        "attach_runtime_agent_endpoint" => Some("agent_endpoint"),
        "apply_runtime_plan" => Some("runtime_plan"),
        "run_setup_script" => Some("setup_script"),
        "start_egress_proxy" => Some("egress"),
        "start_runtime_adapters" => Some("runtime_adapters"),
        "start_runtime_processes" => Some("runtime_processes"),
        "start_tunnel_session" | "stop_tunnel_session" | "attach_runtime_environment" => {
            Some("operation_stream")
        }
        "wait_storage_attach" => Some("storage_attach"),
        "ready" => Some("ready"),
        phase if phase.starts_with("stop_tunnel_session_") => Some("operation_stream"),
        phase if phase.starts_with("stop_egress_proxy") => Some("teardown"),
        _ => None,
    }
}

pub(super) fn lifecycle_attributes(payload: &Value) -> Value {
    let mut attributes = Map::new();
    if let Some(object) = payload.as_object() {
        for (key, value) in object {
            if matches!(
                key.as_str(),
                "timestamp" | "level" | "event" | "sandboxInstanceId" | "operation"
            ) {
                continue;
            }
            attributes.insert(key.clone(), value.clone());
        }
    }
    Value::Object(attributes)
}

pub(super) fn started_event_name(operation: StartupOperation) -> &'static str {
    match operation {
        StartupOperation::Init => "sandbox_init_started",
        StartupOperation::Resume => "sandbox_resume_started",
    }
}

pub(super) fn phase_failed_event_name(operation: StartupOperation) -> &'static str {
    match operation {
        StartupOperation::Init => "sandbox_init_phase_failed",
        StartupOperation::Resume => "sandbox_resume_phase_failed",
    }
}

pub(super) fn phase_started_event_name(operation: StartupOperation) -> &'static str {
    match operation {
        StartupOperation::Init => "sandbox_init_phase_started",
        StartupOperation::Resume => "sandbox_resume_phase_started",
    }
}

pub(super) fn phase_completed_event_name(operation: StartupOperation) -> &'static str {
    match operation {
        StartupOperation::Init => "sandbox_init_phase_completed",
        StartupOperation::Resume => "sandbox_resume_phase_completed",
    }
}

pub(super) fn failed_event_name(operation: StartupOperation) -> &'static str {
    match operation {
        StartupOperation::Init => "sandbox_init_failed",
        StartupOperation::Resume => "sandbox_resume_failed",
    }
}

pub(super) fn transcript_event_name(operation: StartupOperation) -> &'static str {
    match operation {
        StartupOperation::Init => "sandbox_init_transcript",
        StartupOperation::Resume => "sandbox_resume_transcript",
    }
}

File: packages/sandboxd/startupdiagnostics/operation_records.go
package startupdiagnostics

import (
	"encoding/json"
	"fmt"
	"strings"

	"github.com/mistle/sandboxd/protocol"
)

type ActivationOperation struct {
	OperationKind protocol.ActivationOperationKind
}

func OperationRecordLine(operation ActivationOperation, observedAt string, event string, payload map[string]any) (*string, error) {
	eventNames, err := operationEventNames(operation)
	if err != nil {
		return nil, err
	}
	var record map[string]any
	switch {
	case event == eventNames.Started, event == eventNames.Failed:
		return nil, nil
	case event == eventNames.PhaseStarted:
		phase, ok := operationRecordPhase(payload)
		if !ok {
			return nil, nil
		}
		record = lifecycleOperationRecord(observedAt, phase, "started", payload)
	case event == eventNames.PhaseCompleted:
		phase, ok := operationRecordPhase(payload)
		if !ok {
			return nil, nil
		}
		record = lifecycleOperationRecord(observedAt, phase, "completed", payload)
	case event == eventNames.PhaseFailed:
		phase, ok := operationRecordPhase(payload)
		if !ok {
			return nil, nil
		}
		record = lifecycleOperationRecord(observedAt, phase, "failed", payload)
	case event == eventNames.Transcript:
		record = map[string]any{
			"kind":          "transcript",
			"observedAt":    observedAt,
			"phase":         optionalOperationLifecyclePhase(payload),
			"source":        "sandboxd",
			"stream":        stringValue(payload, "stream", "system"),
			"payloadBase64": stringValue(payload, "payloadBase64", ""),
		}
	default:
		return nil, nil
	}

	serialized, err := json.Marshal(record)
	if err != nil {
		return nil, fmt.Errorf("failed to serialize operation record: %w", err)
	}
	line := string(serialized) + "\n"
	return &line, nil
}

func OperationLifecyclePhase(phase string) (string, bool) {
	switch phase {
	case "apply_git_identity":
		return "git_identity", true
	case "attach_runtime_agent_endpoint":
		return "agent_endpoint", true
	case "apply_runtime_plan":
		return "runtime_plan", true
	case "run_setup_script":
		return "setup_script", true
	case "start_egress_proxy":
		return "egress", true
	case "start_runtime_adapters":
		return "runtime_adapters", true
	case "start_runtime_processes":
		return "runtime_processes", true
	case "start_tunnel_session", "stop_tunnel_session", "attach_runtime_environment":
		return "operation_stream", true
	case "ready":
		return "ready", true
	default:
		if strings.HasPrefix(phase, "stop_tunnel_session_") {
			return "operation_stream", true
		}
		if strings.HasPrefix(phase, "stop_egress_proxy") {
			return "teardown", true
		}
		return "", false
	}
}

func LifecycleAttributes(payload map[string]any) map[string]any {
	attributes := make(map[string]any)
	for key, value := range payload {
		switch key {
		case "timestamp", "level", "event", "sandboxInstanceId", "operation":
			continue
		default:
			attributes[key] = value
		}
	}
	return attributes
}

func lifecycleOperationRecord(observedAt string, phase string, status string, payload map[string]any) map[string]any {
	return map[string]any{
		"kind":       "lifecycle",
		"observedAt": observedAt,
		"phase":      phase,
		"status":     status,
		"source":     "sandboxd",
		"message":    fmt.Sprintf("%s %s", phase, status),
		"attributes": LifecycleAttributes(payload),
	}
}

func operationRecordPhase(payload map[string]any) (string, bool) {
	rawPhase, ok := payload["phase"].(string)
	if !ok {
		return "", false
	}
	return OperationLifecyclePhase(rawPhase)
}

func optionalOperationLifecyclePhase(payload map[string]any) any {
	rawPhase, ok := payload["phase"].(string)
	if !ok {
		return nil
	}
	phase, ok := OperationLifecyclePhase(rawPhase)
	if !ok {
		return nil
	}
	return phase
}

func stringValue(payload map[string]any, key string, defaultValue string) string {
	value, ok := payload[key].(string)
	if !ok {
		return defaultValue
	}
	return value
}

type activationOperationEventNames struct {
	Started        string
	PhaseStarted   string
	PhaseCompleted string
	PhaseFailed    string
	Failed         string
	Transcript     string
}

func operationEventNames(operation ActivationOperation) (activationOperationEventNames, error) {
	prefix, err := activationOperationEventPrefix(operation)
	if err != nil {
		return activationOperationEventNames{}, err
	}
	return activationOperationEventNames{
		Started:        prefix + "_started",
		PhaseStarted:   prefix + "_phase_started",
		PhaseCompleted: prefix + "_phase_completed",
		PhaseFailed:    prefix + "_phase_failed",
		Failed:         prefix + "_failed",
		Transcript:     prefix + "_transcript",
	}, nil
}

func activationOperationEventPrefix(operation ActivationOperation) (string, error) {
	switch operation.OperationKind {
	case protocol.ActivationOperationStart:
		return "sandbox_start", nil
	case protocol.ActivationOperationResume:
		return "sandbox_resume", nil
	case protocol.ActivationOperationSetupCheck:
		return "sandbox_setup_check", nil
	case protocol.ActivationOperationSnapshot:
		return "sandbox_snapshot", nil
	default:
		return "", fmt.Errorf("unsupported activation operation kind: %s", operation.OperationKind)
	}
}
